quixstreams/sources/community/file/base.py

```diff
@@ -0,0 +1,216 @@
+import logging
+from abc import abstractmethod
+from pathlib import Path
+from time import sleep
+from typing import BinaryIO, Callable, Iterable, Optional, Union
+
+from quixstreams.models import Topic, TopicConfig
+from quixstreams.sources import (
+    ClientConnectFailureCallback,
+    ClientConnectSuccessCallback,
+    Source,
+)
+
+from .components import FileFetcher, raw_filestream_deserializer
+from .compressions import CompressionName
+from .formats import Format, FormatName
+
+__all__ = ("FileSource",)
+
+logger = logging.getLogger(__name__)
+
+
+class FileSource(Source):
+    """
+    An interface for extracting records using a file-based client.
+
+    It recursively iterates from a provided path (file or folder) and
+    processes all found files by parsing and producing the given records contained
+    in each file as individual messages to a kafka topic.
+
+    Requires defining methods for navigating folders and retrieving/opening raw
+    files for the respective client.
+
+    When these abstract methods are defined, a FileSource will be able to:
+    1. Prepare a list of files to download, and retrieve them sequentially
+    2. Retrieve file contents asynchronously by downloading the upcoming one in the
+        background
+    3. Decompress and deserialize the current file to loop through its records
+    4. Apply a replay delay for each contained record based on previous record
+    5. Serialize and produce respective messages to Kafka based on provided `setters`
+    """
+
+    def __init__(
+        self,
+        filepath: Union[str, Path],
+        key_setter: Optional[Callable[[object], object]] = None,
+        value_setter: Optional[Callable[[object], object]] = None,
+        timestamp_setter: Optional[Callable[[object], int]] = None,
+        file_format: Union[Format, FormatName] = "json",
+        compression: Optional[CompressionName] = None,
+        has_partition_folders: bool = False,
+        replay_speed: float = 1.0,
+        name: Optional[str] = None,
+        shutdown_timeout: float = 30,
+        on_client_connect_success: Optional[ClientConnectSuccessCallback] = None,
+        on_client_connect_failure: Optional[ClientConnectFailureCallback] = None,
```

Comment:
If those arguments are not overloaded, we could just catch and pass kwargs, WDYT?
```suggestion
        **kwargs
```

PS. if you agree, remember to remove those from the docstring as well